### PR TITLE
Kpod logs

### PR DESCRIPTION
**Pull request by 14rcole**
Add `kpod logs` command for viewing the logs of a container

### CODE REVIEW

File: cmd/kpod/logs.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+			Usage: "Follow log output.  The default is false",
+		},
+		cli.StringFlag{
+			Name:  "since",
+			Usage: "show logs since TIMESTAMP",
+		},
+		cli.StringFlag{
+			Name:  "tail",
+			Usage: "output the specified number of LINES at the end of the logs.  Defaults to 'all'",
+		},
```
Comment (TomSweeneyRedHat) [inline cmd/kpod/logs.go]: nit of a nit.  Some of these usage statements start with an uppercase, others with lowercase.  I'd prefer upper, either is fine, just be consistent plz.

File: docs/kpod-logs.1.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+## OPTIONS
+
+**--details**
+
+Show extra details provided to the logs
+
+**--follow, -f**
+
+Follow log output.  Default is falose
```
Comment (TomSweeneyRedHat) [inline docs/kpod-logs.1.md]: falose to false

**Comment by TomSweeneyRedHat** [comment]
couple nits.  Any tests inbound?

File: libkpod/logs.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+// LogOptions contains all of the options for displaying logs in kpod
+type LogOptions struct {
+	Details   bool
+	Follow    bool
+	SinceTime time.Time
+	Tail      int64
+}
+
+// GetLogs gets each line of a log file and, if it matches the criteria in logOptions, sends it down logChan
+func GetLogs(server *ContainerServer, container string, logChan chan string, opts LogOptions) error {
```
Comment (mheon) [inline libkpod/logs.go]: nit: Can we make this func (server *ContainerServer) GetLogs(...) instead? I'd prefer to make our public API methods of ContainerServer to the greatest extent possible.

Comment (14rcole) [inline libkpod/logs.go]: I can do that

**Comment by 14rcole** [comment]
@TomSweeneyRedHat tests will be coming, I have at least one PR that this depends on for those tests to pass though, so I wanted to get the PR started first

**Comment by TomSweeneyRedHat** [comment]
LGTM, assuming happy ci.

File: cmd/kpod/logs.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	}
+	logsDescription = "The kpod logs command batch-retrieves whatever logs are present for a container at the time of execution.  This does not guarantee execution" +
+		"order when combined with kpod run (i.e. your run may not have generated any logs at the time you execute kpod logs"
+	logsCommand = cli.Command{
+		Name:        "logs",
+		Usage:       "Fetch the logs of a container",
+		Description: logsDescription,
+		Flags:       logsFlags,
+		Action:      logsCmd,
+		ArgsUsage:   "",
```
Comment (rhatdan) [inline cmd/kpod/logs.go]: What are the args?

File: cmd/kpod/logs.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	}
+)
+
+func logsCmd(c *cli.Context) error {
+	args := c.Args()
+	if len(args) != 1 {
+		return errors.Errorf("'kpod logs' requires exactly one argument")
```
Comment (rhatdan) [inline cmd/kpod/logs.go]: What args?

File: cmd/kpod/logs.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	opts.SinceTime = time.Time{}
+	if c.IsSet("since") {
+		// parse time, error out if something is wrong
+	}
+	opts.Tail = -1
+	if c.IsSet("tail") {
+		// parse tail: it needs to be all or a positive number
+		tailStr := c.String("tail")
+
+		if tailStr != "all" {
```
Comment (rhatdan) [inline cmd/kpod/logs.go]: Is it valid to just do a
docker logs --tail CONTAINER

File: docs/kpod-logs.1.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+## SYNOPSIS
+**kpod** **logs** [*options* [...]] container
+
+## DESCRIPTION
+The kpod logs command batch-retrieves whatever logs are present for a container at the time of execution.  This does not guarantee execution order when combined with kpod run (i.e. your run may not have generated any logs at the time you execute kpod logs
+
+## OPTIONS
+
+**--details**
```
Comment (rhatdan) [inline docs/kpod-logs.1.md]: Why is this hidden above and exposed here. If it is hidden then it should not be in the man page.

File: docs/kpod-logs.1.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+Follow log output.  Default is false
+
+**--since=TIMESTAMP**
+
+Show logs since TIMESTAMP
+
+**--tail=LINES**
+
+Ouput the specified number of LINES at the end of the logs.  Defaults to 'all'
```
Comment (rhatdan) [inline docs/kpod-logs.1.md]: Not sure how all makes sense here unless --tail without an option is valid.

File: docs/kpod-logs.1.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+## EXAMPLE
+
+kpod logs b3f2436bdb978c1d33b1387afb5d7ba7e3243ed2ce908db431ac0069da86cb45
+
```
Comment (rhatdan) [inline docs/kpod-logs.1.md]: How about some example of the logs.

File: libkpod/logs.go [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+	"path"
+	"strings"
+	"time"
+
+	"github.com/hpcloud/tail"
+)
+
+const (
+	// LogsBaseDir specifies the default directory in which logs are stored
+	logsBaseDir = "/var/log/containers"
```
Comment (rhatdan) [inline libkpod/logs.go]: Kpod config has

./libkpod/config.go:			LogDir:         "/var/log/crio/pods",

File: cmd/kpod/logs.go (hunk 6 above)
Comment (14rcole) [inline cmd/kpod/logs.go]: Well in that case, docker logs would interpret CONTAINER as the parameter for tail and given an error that docker logs requires at least one argument, but `docker logs CONTAINER --tail` should return an error.  I'll fix this and repushed

File: docs/kpod-logs.1.md (hunk 7 above)
Comment (14rcole) [inline docs/kpod-logs.1.md]: I forgot that this needed to be hidden until the `--log-opt` flag could be implemented in `kpod create`.  I'll remove it from the man page for now

File: docs/kpod-logs.1.md (hunk 9 above)
Comment (14rcole) [inline docs/kpod-logs.1.md]: I'll put that in as soon as it's working, which is dependent on #713

File: libkpod/logs.go (hunk 10 above)
Comment (14rcole) [inline libkpod/logs.go]: That needs to be removed,  kpod logs should read the log dir from the config.  I'll fix that

**Comment by rhatdan** [comment]
@14rcole Needs rebase.

File: libkpod/container_server.go [hunk 11]
```diff
@@ -161,6 +161,7 @@ func (c *ContainerServer) Update() error {
 		logrus.Warnf("could not read containers and sandboxes: %v", err)
 		return err
 	}
+	fmt.Println("Containers in store: ", len(containers))
```
Comment (mrunalp) [inline libkpod/container_server.go]: This is probably unintentional.

File: docs/kpod-logs.1.md [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+**--tail=LINES**
+
+Ouput the specified number of LINES at the end of the logs.  LINES must be a positive integer or 'all'.  Defaults to 'all'
+
+## EXAMPLE
+
+kpod logs b3f2436bdb978c1d33b1387afb5d7ba7e3243ed2ce908db431ac0069da86cb45
+
+2017/08/07 10:16:21 Seeked /var/log/crio/pods/eb296bd56fab164d4d3cc46e5776b54414af3bf543d138746b25832c816b933b/c49f49788da14f776b7aa93fb97a2a71f9912f4e5a3e30397fca7dfe0ee0367b.log - &{Offset:0 Whence:0}
+2017-08-07T10:10:09.055123874-04:00 stdout 1:C 07 Aug 14:10:09.055 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
```
Comment (mrunalp) [inline docs/kpod-logs.1.md]: We should strip out `2017-08-07T10:10:09.055123874-04:00 stdout`. This is the CRI format requirement. Typically applications will add their own time stamp which should be visible.

**Comment by rhatdan** [comment]
@mrunalp @runcom PTAL
Looks like it passed the E2E tests, well at least the ones that ran.

File: cmd/kpod/logs.go [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+		}
+		opts.SinceTime = since
+	}
+	opts.Tail = -1
+	if c.IsSet("tail") {
+		// parse tail: it needs to be all or a positive number
+		tailStr := c.String("tail")
+		if tailStr == "" {
+			return errors.Errorf("flag needs an argument: --tail")
+		} else if tailStr != "all" {
```
Comment (rhatdan) [inline cmd/kpod/logs.go]: Is anyone ever going to specify all?  I think leaving this as an unsigned int would be fine.

File: cmd/kpod/logs.go [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+	if c.IsSet("since") {
+		// parse time, error out if something is wrong
+		since, err := time.Parse("2006-01-02T15:04:05.999999999-07:00", c.String("since"))
+		if err != nil {
+			return errors.Wrapf(err, "could not parse time: %q", c.String("since"))
+		}
+		opts.SinceTime = since
+	}
+	opts.Tail = -1
+	if c.IsSet("tail") {
```
Comment (rhatdan) [inline cmd/kpod/logs.go]: Lets just make Tail and int.

File: cmd/kpod/logs.go [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+		},
+		cli.BoolFlag{
+			Name:  "follow, f",
+			Usage: "Follow log output.  The default is false",
+		},
+		cli.StringFlag{
+			Name:  "since",
+			Usage: "Show logs since TIMESTAMP",
+		},
+		cli.StringFlag{
```
Comment (rhatdan) [inline cmd/kpod/logs.go]: IntFlag

File: cmd/kpod/main.go [hunk 16]
```diff
@@ -31,6 +31,8 @@ func main() {
 		inspectCommand,
 		loadCommand,
 		mountCommand,
+		loadCommand,
```
Comment (rhatdan) [inline cmd/kpod/main.go]: This looks like a mistake.

File: completions/bash/kpod [hunk 17]
```diff
@@ -80,6 +84,28 @@ _kpod_launch() {
      _complete_ "$options_with_args" "$boolean_options"
 }
 
+_kpod_logs() {
+     local options_with_args="
+     --since
+     --tail
+     "
+     local boolean_options="
+     --json
```
Comment (rhatdan) [inline completions/bash/kpod]: --json is not a valid option

File: docs/kpod-logs.1.md [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+
+Follow log output.  Default is false
+
+**--since=TIMESTAMP**
+
+Show logs since TIMESTAMP
+
+**--tail=LINES**
+
+Ouput the specified number of LINES at the end of the logs.  LINES must be a positive integer or 'all'.  Defaults to 'all'
```
Comment (rhatdan) [inline docs/kpod-logs.1.md]: Remove the "or 'all'"

File: libkpod/logs.go [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+func logSinceTime(sinceTime time.Time, logStr string) (bool, error) {
+	timestamp := strings.Split(logStr, " ")[0]
+	logTime, err := time.Parse("2006-01-02T15:04:05.999999999-07:00", timestamp)
+	if err != nil {
+		return false, err
+	}
+	return logTime.After(sinceTime) || logTime.Equal(sinceTime), nil
+}
+
+func secondSpaceIndex(line string) int {
```
Comment (rhatdan) [inline libkpod/logs.go]: Could you give a short description of what this function does?

File: cmd/kpod/logs.go (hunk 15 above)
Comment (14rcole) [inline cmd/kpod/logs.go]: The value can be "all", do you want me to remove that option?

File: cmd/kpod/logs.go (hunk 13 above)
Comment (14rcole) [inline cmd/kpod/logs.go]: Works for me

**Comment by rhatdan** [comment]
LGTM
@mrunalp @runcom PTAL
Lets get this merged.

**Comment by mrunalp** [comment]
LGTM

**Pull request merged**